external_apps/bid/bid_action/owner_decision.go

```diff
@@ -0,0 +1,200 @@
+	}
+	bidConv, err := bidMasterStore.BidConv.WithPrefixType(bid_data.BidStateActive).Get(ownerDecision.BidConvId)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrGettingBidConv, ownerDecision.Tags(), err)
+	}
+
+	//2. check expiry
+	deadLine := time.Unix(bidConv.DeadlineUTC, 0)
+
+	if deadLine.Before(ctx.Header.Time.UTC()) {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrExpiredBid, ownerDecision.Tags(), err)
+	}
+
+	//3. check asset availability
+	available, err := IsAssetAvailable(ctx, bidConv.AssetName, bidConv.AssetType, bidConv.AssetOwner)
+	if err != nil || available == false {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrInvalidAsset, ownerDecision.Tags(), err)
+	}
+
+	//4. check owner's identity
+	if !ownerDecision.Owner.Equal(bidConv.AssetOwner) {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrWrongAssetOwner, ownerDecision.Tags(), err)
+	}
+
+	//5. get active bid offer
+	activeOffer, err := bidMasterStore.BidOffer.GetActiveOffer(ownerDecision.BidConvId, bid_data.TypeBidOffer)
+	// in this case, there must be an existing active offer
+	if err != nil || activeOffer == nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrGettingActiveOffer, ownerDecision.Tags(), err)
+	}
+
+	//6. if reject
+	if ownerDecision.Decision != bid_data.RejectBid && ownerDecision.Decision != bid_data.AcceptBid {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrInvalidOwnerDecision, ownerDecision.Tags(), err)
+	} else if ownerDecision.Decision == bid_data.RejectBid {
+		// deactivate offer and unlock amount
+		err = DeactivateOffer(false, bidConv.Bidder, ctx, activeOffer, bidMasterStore)
+		if err != nil {
+			return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrDeactivateOffer, ownerDecision.Tags(), err)
+		}
+		// close bid conversation
+		err = CloseBidConv(bidConv, bidMasterStore, bid_data.BidStateRejected)
+		if err != nil {
+			return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrCloseBidConv, ownerDecision.Tags(), err)
+		}
+
+		return helpers.LogAndReturnTrue(ctx.Logger, ownerDecision.Tags(), "owner_reject_bid_success")
+
+	}
+
+	//7. add the amount to owner, in this case offer amount is already being locked from bidder
+	activeOfferCoin := activeOffer.Amount.ToCoin(ctx.Currencies)
+	err = ctx.Balances.AddToAddress(bidConv.AssetOwner.Bytes(), activeOfferCoin)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrAdddingAmountToOwner, ownerDecision.Tags(), err)
+	}
+
+	//8. change offer status to inactive and add it back to bid offer store
+	err = DeactivateOffer(true, bidConv.Bidder, ctx, activeOffer, bidMasterStore)
+	if err != nil {
+		return helpers.LogAndReturnFalse(ctx.Logger, bid_data.ErrDeactivateOffer, ownerDecision.Tags(), err)
+	}
+
+	//9. close the bid conversation
+	err = CloseBidConv(bidConv, bidMasterStore, bid_data.BidStateSucceed)
```

Comment:
If we have multiple conv for one asset , The other conversations would just expire eventually ?

Reply:
and actually if we do not reject those other offers we are still fine.
1. they will be expired
2. if we make counter offers to two people(by design this is not recommended), the first one that deals will get the asset. and the second one will fail to agree to the deal, because the asset right now is no longer valid, since the ownership has changed, and if will fail at asset validation part.